tests/bot/utils/test_redis_cache.py

```diff
@@ -0,0 +1,239 @@
+        """Test that we can get the correct .length from the RedisDict."""
+        await self.redis.clear()
+        await self.redis.set('one', 1)
+        await self.redis.set('two', 2)
+        await self.redis.set('three', 3)
+        self.assertEqual(await self.redis.length(), 3)
+
+        await self.redis.set('four', 4)
+        self.assertEqual(await self.redis.length(), 4)
+
+    async def test_to_dict(self):
+        """Test that the .to_dict method returns a workable dictionary copy."""
+        copy = await self.redis.to_dict()
+        local_copy = {key: value for key, value in await self.redis.items()}
+        self.assertIs(type(copy), dict)
+        self.assertDictEqual(copy, local_copy)
+
+    async def test_clear(self):
+        """Test that the .clear method removes the entire hash."""
+        await self.redis.clear()
+        await self.redis.set('teddy', 'with me')
+        await self.redis.set('in my dreams', 'you have a weird hat')
+        self.assertEqual(await self.redis.length(), 2)
+
+        await self.redis.clear()
+        self.assertEqual(await self.redis.length(), 0)
+
+    async def test_pop(self):
+        """Test that we can .pop an item from the RedisDict."""
+        await self.redis.clear()
+        await self.redis.set('john', 'was afraid')
+
+        self.assertEqual(await self.redis.pop('john'), 'was afraid')
+        self.assertEqual(await self.redis.pop('pete', 'breakneck'), 'breakneck')
+        self.assertEqual(await self.redis.length(), 0)
+
+    async def test_update(self):
+        """Test that we can .update the RedisDict with multiple items."""
+        await self.redis.clear()
+        await self.redis.set("reckfried", "lona")
+        await self.redis.set("bel air", "prince")
+        await self.redis.update({
+            "reckfried": "jona",
+            "mega": "hungry, though",
+        })
+
+        result = {
+            "reckfried": "jona",
+            "bel air": "prince",
+            "mega": "hungry, though",
+        }
+        self.assertDictEqual(await self.redis.to_dict(), result)
+
+    def test_typestring_conversion(self):
+        """Test the typestring-related helper functions."""
+        conversion_tests = (
+            (12, "i|12"),
+            (12.4, "f|12.4"),
+            ("cowabunga", "s|cowabunga"),
+        )
+
+        # Test conversion to typestring
+        for _input, expected in conversion_tests:
+            self.assertEqual(self.redis._to_typestring(_input), expected)
+
+        # Test conversion from typestrings
+        for _input, expected in conversion_tests:
+            self.assertEqual(self.redis._from_typestring(expected), _input)
+
+        # Test that exceptions are raised on invalid input
+        with self.assertRaises(TypeError):
+            self.redis._to_typestring(["internet"])
+            self.redis._from_typestring("o|firedog")
+
+    async def test_increment_decrement(self):
+        """Test .increment and .decrement methods."""
```

Comment:
I would suggest also having a `test_increment_concurrent()` to ensure it's actually safe from race conditions. It *should be* okay with the current lock implementation, but to actually test for it I'd advise something like the following: 
```py
await self.redis.set("test_key", 0)
tasks = []
# Feel free to adjust the count if needed
for _ in range(100):
    task = asyncio.create_task(
        self.redis.increment("test_key", 1))
    tasks.append(task)

await asyncio.gather(*tasks)
val = await self.redis.get("test_key")
self.assertEqual(val, 100)
```

Reply:
oh yeah, I really like this. I'll set that up now. We can probably go even higher than 100, I doubt it'll add much test time if we go for 1000, and I'd feel a bit safer with that.

Reply:
Awesome! Yeah, 1000 or so should more fully cover your bases; I just had no idea what the actual runtime was like so I used 100 as a more conservative estimate. I'm glad the example was helpful though. :-)